nowcasting_dataset/data_sources/sun/raw_data_load_save.py

```diff
@@ -0,0 +1,181 @@
+import datetime
+import time
+import io
+import logging
+from concurrent import futures
+from typing import List, Union, Optional
+
+import fsspec
+import numcodecs
+import pandas as pd
+from tqdm import tqdm
+import xarray as xr
+import numpy as np
+from pathlib import Path
+
+from nowcasting_dataset import geospatial
+
+logger = logging.getLogger(__name__)
+
+
+def get_azimuth_and_elevation(
+    datestamps: List[datetime.datetime], x_centers: List[int], y_centers: List[int]
+) -> (pd.DataFrame, pd.DataFrame):
+    """
+
+    Get Azimuth and elevation positions of the sun
+
+    For a list of datestamps and a list of coordinates, get the azimuth and elevation degrees.
+    Note that the degrees are rounded to 2 decimal places, as we at most need that.
+
+    Args:
+        datestamps: list of datestamps that are needed
+        x_centers: list of x coordinates - ref. OSGB
+        y_centers: list of y coordinates - ref. OSGB
+
+    Returns: Tuple of dataframes for azimuth and elevation.
+        The index is timestamps, and the columns are the x and y coordinates
+
+    """
+
+    logger.debug(
+        f"Will be calculating for {len(datestamps)} datestamps and {len(x_centers)} locations"
+    )
+
+    # create array of index datetime, columns of system_id for both azimuth and elevation
+    azimuth = []
+    elevation = []
+
+    t = time.time()
+    names = []
+    # loop over locations and find azimuth and elevation angles,
+    with futures.ThreadPoolExecutor() as executor:
+
+        logger.debug("Setting up jobs")
+
+        # Submit tasks to the executor.
+        future_azimuth_and_elevation_per_location = []
+        for i in tqdm(range(len(x_centers))):
+
+            name = x_y_to_name(x_centers[i], y_centers[i])
+            if name not in names:
+
+                lat, lon = geospatial.osgb_to_lat_lon(x=x_centers[i], y=y_centers[i])
+
+                future_azimuth_and_elevation = executor.submit(
+                    geospatial.calculate_azimuth_and_elevation_angle,
+                    latitude=lat,
+                    longitude=lon,
+                    datestamps=datestamps,
+                )
+                future_azimuth_and_elevation_per_location.append(
+                    [future_azimuth_and_elevation, name]
+                )
+                names.append(name)
+
+        logger.debug(f"Getting results")
+
+        # Collect results from each thread.
+        for i in tqdm(range(len(future_azimuth_and_elevation_per_location))):
```

Comment:
It might be worth considering changing this line to something like:

```suggestion
        for future_azimuth_and_elevation, name in tqdm(future_azimuth_and_elevation_per_location):
```

Maybe that's more explicit about what this loop is for?  Although it's also quite a long line!

Reply:
yea that looks good